signal-cli 0.8.1 (with libsignal-client now built from src)

mtigas opened the pull request:
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
  - this should supersede #73002 I think
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Took a stab at this after seeing #73002

I [previously had a gist for bundling a libzkgroup.dylib prebuilt binary](https://gist.github.com/mtigas/e4157b7bdbb77e87506ceff2c8f19826), and I've basically done that here with `libzkgroup` as a resource in the formula. `libsignal-client` also has to be built from source now.

Took a bit of trial-and-error since it appears that [libsignal-client](https://github.com/signalapp/libsignal-client/) doesn't like to build with brew-installed `rust` (since `cargo` in that case doesn't fetch the necessary specific version of `rustc` that libsignal-client relies on).

Caveats: this is now marked as a MacOS-only, x86_64-only formula, as the [prebuilt libskgroup.dylib library](https://github.com/signalapp/zkgroup/releases) only contains x86_64 (and b/c I have not looked into using the `.so` for Linux and whether the build steps for libsignal-client work on homebrew-on-linux).

```
$ file /tmp/libzkgroup.dylib
/tmp/libzkgroup.dylib: Mach-O 64-bit dynamically linked shared library x86_64
```

carlocab reviewed (commented):
Thanks for this, @mtigas! I'd been meaning to get to this, but my todo list has been growing increasingly long these days.

mtigas commented:
Updated branch to also build `libzkgroup` from source, added `on_macos`/`on_linux` where I think they're needed; removed `depends_on arch: :x86_64` and `depends_on :macos`.

carlocab reviewed (commented):
Had a closer look now. Just a few suggestions.

carlocab commented on Formula/signal-cli.rb:
```diff
@@ -1,17 +1,89 @@
+      # build
+      on_macos do
+        system "make", "mac_dylib"
+        #system "make", "libzkgroup" # above is deprecated and will go away in a future version
```
```suggestion
        # system "make", "libzkgroup" # above is deprecated and will go away in a future version
```
`brew` is picky about formatting. This should fix the CI failure.

mtigas commented on Formula/signal-cli.rb:
```diff
@@ -1,17 +1,89 @@
+
+      # build
+      Dir.chdir("java")
+      system "gsed", "-i", "s/, ':android'//", "settings.gradle"
```
thanks! couldn't figure out the clean way to do this (given the [divergent behavior](https://izziswift.com/sed-i-command-for-in-place-editing-to-work-with-both-gnu-sed-and-bsd-osx/) of `-i` in bsd vs gnu sed)

carlocab commented on Formula/signal-cli.rb:
```diff
@@ -1,17 +1,84 @@
+      on_linux do
+        system "make", "libzkgroup"
+        cd "ffi/java/src/main/resources" do
+            system "zip", "-u", "#{libexec}/lib/zkgroup-java-0.7.0.jar", "./libzkgroup.so"
```
```suggestion
          system "zip", "-u", "#{libexec}/lib/zkgroup-java-0.7.0.jar", "./libzkgroup.so"
```
Like I mentioned, `brew` is very particular about formatting 😅 

To be sure, you can always run
```
brew audit --fix signal-cli
```
or
```
brew style --fix signal-cli
```
locally to have it fix any formatting issues it doesn't like.

mtigas replied:
ah! forgot to run that this time. fixing.

SMillerDev commented on Formula/signal-cli.rb:
```diff
@@ -1,17 +1,84 @@
 
   depends_on "openjdk"
 
+  resource "libsignal-client" do
+    # per https://github.com/AsamK/signal-cli/wiki/Provide-native-lib-for-libsignal#libsignal-client
+    # we want the specific libsignal-client version from 'signal-cli-0.8.1/lib/signal-client-XXXX-X.X.X.jar'
+    version "java-0.2.3"
+    url "https://github.com/signalapp/libsignal-client/archive/refs/tags/#{version}.tar.gz"
+    sha256 "730c1dc113da5227920716656d8f888e1af167208e095a8cac3de9c0d83890c4"
+  end
+
+  resource "libzkgroup" do
+    # per https://github.com/AsamK/signal-cli/wiki/Provide-native-lib-for-libsignal#libzkgroup
+    # we want the latest release version
+    url "https://github.com/signalapp/zkgroup/archive/refs/tags/v0.7.2.tar.gz"
+    sha256 "fdd03bbf584533963d1be40ab238d4e6199b379e8112f6aaf5cd9493b7f1fb47"
+  end
```
Can't these be formulae?

carlocab replied:
Would that be useful if they're not needed in any other formulae?

SMillerDev replied:
Yeah, if it's independent software we should try and make it a formula. That way it can be checked for updates and everything.

carlocab replied:
I guess `libzkgroup` can be its own formula, since we always want the latest version. `libsignal-client` seems to have specific version requirements, but I don't mind having a formula as the latest version to see if it works.

Though I guess it would be annoying to have a `libsignal-client` formula that isn't actually useful for `signal-cli`.

mtigas replied:
Hm. Let me see what things look like if I just try to entirely build signal-cli from source?

Right now I'm just following the upstream [recommended install steps](https://github.com/AsamK/signal-cli#installation) which unfortunately is basically "download the prebuilt binary package, and if you're not on linux-x86_64 you also have to rebuild/replace the libsignal & libzkgroup libraries inside that prebuilt package."

(It looks like [0.8.0](https://github.com/AsamK/signal-cli/blob/master/CHANGELOG.md#080---2021-02-14) added the libsignal-client dependency, and [0.7.0](https://github.com/AsamK/signal-cli/blob/master/CHANGELOG.md#070---2020-12-15) added the libzkgroup dependency, and the distribution only includes them for linux x86_64.)

While technically dynamic libraries, I'm not 100% sure how externalizing them would work within the Java packaging and whatnot? (I am not a Java dev, so all I know is that these libs end up getting zipped into the .jar files that this thing needs. Which I think precludes them from being symlinked/etc the way dynamic libraries normally are? Again, not a Java dev.)

carlocab replied:
> Hm. Let me see what things look like if I just try to entirely build signal-cli from source?

This sounds worth a try if you don't mind looking into it 👍

patcon replied:
My understanding is the same as @mtigas, and that separate formulae might not make sense here (due to how install happens directly into jars?)

carlocab commented on Formula/signal-cli.rb:
```diff
@@ -1,17 +1,84 @@
+        inreplace "settings.gradle", ", ':android'", ""
+        system "./build_jni.sh", "desktop"
+        cd "java/src/main/resources" do
+          on_macos do
+            system "zip", "-u", "#{libexec}/lib/signal-client-java-0.2.3.jar", "./libsignal_jni.dylib"
+          end
+          on_linux do
+            system "zip", "-u", "#{libexec}/lib/signal-client-java-0.2.3.jar", "./libsignal_jni.so"
+          end
```
```suggestion
          system "zip", "-u", "#{libexec}/lib/signal-client-java-0.2.3.jar", "./#{shared_library("libsignal_jni")}"
```
We can combine this into one step.

carlocab commented on Formula/signal-cli.rb:
```diff
@@ -1,17 +1,84 @@
+      system "zip", "-d", "#{libexec}/lib/zkgroup-java-0.7.0.jar", "libzkgroup.so"
+
+      # build & embed library for current platform
+      on_macos do
+        system "make", "mac_dylib"
+        # system "make", "libzkgroup" # above is deprecated and will go away in a future version
+        cd "ffi/java/src/main/resources" do
+          system "zip", "-u", "#{libexec}/lib/zkgroup-java-0.7.0.jar", "./libzkgroup.dylib"
+        end
+      end
+      on_linux do
+        system "make", "libzkgroup"
+        cd "ffi/java/src/main/resources" do
+          system "zip", "-u", "#{libexec}/lib/zkgroup-java-0.7.0.jar", "./libzkgroup.so"
+        end
+      end
```
```suggestion
      target = "mac_dylib"
      on_linux do
        target = "libzkgroup"
      end
      
      system "make", target
      # system "make", "libzkgroup" # above is deprecated and will go away in a future version
      cd "ffi/java/src/main/resources" do
        system "zip", "-u", "#{libexec}/lib/zkgroup-java-0.7.0.jar", "./#{shared_library("libzkgroup")}"
      end
```

patcon commented:
Is there a way to have this formula work "from source", but also offer the direct binary download with the same swapping of .dylib files? So allow ppl to skip building step and all the gradle deps?

carlocab commented:
> Is there a way to have this formula work "from source", but also offer the direct binary download with the same swapping of .dylib files? So allow ppl to skip building step and all the gradle deps?

When we build this from source we also use that source build to build a bottle (pre-built binary package). Most users will install from the bottle, so they're already skipping the build and gradle deps.

patcon commented:
Wanted to confirm that this works for me, and I'm using it locally :)

carlocab reviewed (approved):
Rebased to make sure this still builds after all the updates to the dependencies. If it does, LGTM. Thanks, @mtigas, and thanks for your patience here.
